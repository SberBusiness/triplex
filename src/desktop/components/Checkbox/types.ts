import React from 'react';

/** Свойства компонента Checkbox. */
export interface ICheckboxProps extends Omit<React.InputHTMLAttributes<HTMLInputElement>, 'type'> {
    /** Объект label-атрибутов. */
    labelAttributes?: React.LabelHTMLAttributes<HTMLLabelElement>;
    /** Признак частичного типа выбора. */
    bulk?: boolean;
}
