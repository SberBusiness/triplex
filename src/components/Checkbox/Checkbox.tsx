import React, {useRef, useState} from 'react';
import {ICheckboxProps} from '@sberbusiness/triplex/components/Checkbox/types';
import {Text} from '@sberbusiness/triplex/components/Typography/Text';
import {ELineType, ETextSize} from '@sberbusiness/triplex/components/Typography/enums';
import {classnames} from '@sberbusiness/triplex/utils/classnames/classnames';
import {EFocusSource} from '@sberbusiness/triplex/enums/EFocusSource';
import {CheckboxbulkStsIcon16} from '@sberbusiness/icons/CheckboxbulkStsIcon16';
import {CheckboxtickStsIcon16} from '@sberbusiness/icons/CheckboxtickStsIcon16';

/** Чекбокс с описанием. */
export const Checkbox = React.forwardRef<HTMLInputElement, ICheckboxProps>((props, ref) => {
    const {children, className, onFocus, onBlur, disabled, bulk, labelAttributes, ...inputAttributes} = props;
    const [focusVisible, setFocusVisible] = useState(false);
    const focusSource = useRef(EFocusSource.NONE);
    const classNames = classnames('cssClass[checkbox]', className);
    const classNamesLabel = classnames(
        'cssClass[label]',
        {'cssClass[enabled]': !disabled, 'cssClass[nonempty]': !!children},
        labelAttributes?.className
    );
    const inputRef = useRef<HTMLInputElement | null>(null);

    /** Обработчик нажатия мыши. */
    const handleMouseDown = (event: React.MouseEvent<HTMLLabelElement>) => {
        if (!disabled) {
            if (focusSource.current === EFocusSource.NONE) {
                focusSource.current = EFocusSource.MOUSE;
            }
        }
        labelAttributes?.onMouseDown?.(event);
    };

    /** Обработчик клика. */
    const handleClick = (event: React.MouseEvent<HTMLLabelElement>) => {
        if (!disabled) {
            if (event.target === inputRef.current) {
                // Сбрасываем состояние, если событие пришло от чекбокса не в фокусе. (Safari)
                if (event.target !== document.activeElement) {
                    focusSource.current = EFocusSource.NONE;
                }
            } else if (focusSource.current === EFocusSource.NONE) {
                focusSource.current = EFocusSource.MOUSE;
            }
        }
        labelAttributes?.onClick?.(event);
    };

    /** Обработчик получения фокуса. */
    const handleFocus = (event: React.FocusEvent<HTMLInputElement>) => {
        if (focusSource.current !== EFocusSource.MOUSE) {
            focusSource.current = EFocusSource.KEYBOARD;
            setFocusVisible(true);
        } else if (focusVisible) {
            setFocusVisible(false);
        }
        onFocus?.(event);
    };

    /** Обработчик потери фокуса. */
    const handleBlur = (event: React.FocusEvent<HTMLInputElement>) => {
        if (focusSource.current !== EFocusSource.NONE) {
            // При переключении окон/вкладок состояние не сбрасываем.
            if (event.target !== document.activeElement) {
                focusSource.current = EFocusSource.NONE;
                setFocusVisible(false);
            }
        }
        onBlur?.(event);
    };

    /** Функция для хранения ссылки. */
    const setRef = (instance: HTMLInputElement | null) => {
        inputRef.current = instance;
        if (typeof ref === 'function') {
            ref(instance);
        } else if (ref) {
            ref.current = instance;
        }
    };

    /** Отрисовка галочки чекбокса. */
    const renderCheckmarkIcon = () => {
        const className = 'cssClass[checkmarkIcon]';

        return bulk ? <CheckboxbulkStsIcon16 className={className} /> : <CheckboxtickStsIcon16 className={className} />;
    };

    return (
        <Text
            tag="label"
            size={ETextSize.B1}
            line={ELineType.EXTRA}
            htmlFor={inputAttributes.id}
            {...labelAttributes}
            className={classNamesLabel}
            onMouseDown={handleMouseDown}
            onClick={handleClick}
        >
            <input
                type="checkbox"
                className={classNames}
                disabled={disabled}
                onFocus={handleFocus}
                onBlur={handleBlur}
                data-focus-visible={focusVisible ? '' : undefined}
                ref={setRef}
                {...inputAttributes}
            />
            <span className="cssClass[checkboxIcon]" />
            {renderCheckmarkIcon()}
            {children}
        </Text>
    );
});

Checkbox.displayName = 'Checkbox';
