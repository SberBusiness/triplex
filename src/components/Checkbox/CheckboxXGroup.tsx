import React from 'react';
import {TIndentSize} from '@sberbusiness/triplex/consts/IndentConst';
import {cssClass} from '@sberbusiness/triplex/utils/cssClass';
import {classnames} from '@sberbusiness/triplex/utils/classnames/classnames';

/** Свойства компонента CheckboxXGroup. */
export interface ICheckboxXGroupProps extends React.HTMLAttributes<HTMLDivElement> {
    /** Размер отступа. */
    indent?: TIndentSize;
}

/** Группа чекбоксов с направлением по оси X. */
export const CheckboxXGroup: React.FC<ICheckboxXGroupProps> = (props) => {
    const {children, className, indent = 12, ...rest} = props;
    const classNames = classnames('cssClass[checkboxXGroup]', cssClass(`indent-${indent}`), className);

    return (
        <div className={classNames} role="group" {...rest}>
            {children}
        </div>
    );
};

CheckboxXGroup.displayName = 'CheckboxXGroup';
