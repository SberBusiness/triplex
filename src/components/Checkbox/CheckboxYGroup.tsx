import React from 'react';
import {classnames} from '@sberbusiness/triplex/utils/classnames/classnames';

/** Свойства компонента CheckboxYGroup. */
export interface ICheckboxYGroupProps extends React.HTMLAttributes<HTMLDivElement> {}

/** Группа чекбоксов с направлением по оси Y. */
export const CheckboxYGroup: React.FC<ICheckboxYGroupProps> = (props) => {
    const {children, className, ...rest} = props;
    const classNames = classnames('cssClass[checkboxYGroup]', className);

    return (
        <div className={classNames} role="group" {...rest}>
            {children}
        </div>
    );
};

CheckboxYGroup.displayName = 'CheckboxYGroup';
