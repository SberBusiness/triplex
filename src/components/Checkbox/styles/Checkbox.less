@import 'src/styles/style';

.label {
    display: inline-block;
    position: relative;
    min-width: @ufs-theme-size-icon-16;
    min-height: @ufs-theme-size-icon-16;

    &.enabled {
        cursor: pointer;
    }

    &.nonempty {
        padding-left: 24px;
    }

    &[hidden] {
        display: none; // Screenshot tests
    }
}

.checkbox {
    margin: 0;
    opacity: 0;
}

.checkboxIcon {
    box-sizing: border-box;
    background-color: @color-snowy;
    border: 1px solid @color-asphalt-03;
    border-radius: @ufs-theme-base;
}

.checkmarkIcon {
    transform: scale(0);
}

.checkboxIcon,
.checkmarkIcon {
    pointer-events: none;
}

.checkbox,
.checkboxIcon,
.checkmarkIcon {
    position: absolute;
    width: @ufs-theme-size-icon-16;
    height: @ufs-theme-size-icon-16;
    left: 0;

    .nonempty & {
        top: 2px;
    }
}

.checkbox:enabled {
    cursor: pointer;
}

.checkbox:checked {
    + .checkboxIcon {
        background-color: @color-malachite-02;
        border-color: @color-malachite-02;
    }

    ~ .checkmarkIcon {
        transform: scale(1);
        transition: transform 200ms ease-out;
    }
}

.checkbox:hover + .checkboxIcon {
    border-color: @color-malachite-03;
}

.checkbox:disabled {
    + .checkboxIcon {
        background-color: @color-asphalt-05;
        border-color: @color-asphalt-03;
    }

    ~ .checkmarkIcon path {
        fill: @color-asphalt-03;
    }
}

.checkbox[data-focus-visible]:focus + .checkboxIcon {
    border-color: @color-sunny-01;
}

.checkbox:checked:enabled:hover + .checkboxIcon {
    background-color: @color-malachite-03;
}
