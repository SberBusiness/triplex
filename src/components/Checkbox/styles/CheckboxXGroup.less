@import 'src/styles/style';

@indent: 12, 16, 20, 24, 28, 32;

.checkboxXGroup {
    display: inline-flex;
    flex-flow: row wrap;

    each(@indent, {
        &.indent-@{value} {
            margin-left: unit(-@value / 2, px);
            margin-right: unit(-@value / 2, px);

            .label {
                margin-left: unit(@value / 2, px);
                margin-right: unit(@value / 2, px);
            }
        }
    });
}
