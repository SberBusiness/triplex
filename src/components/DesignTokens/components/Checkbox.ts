import {TDesignTokenValue, TDesignTokenValues} from '@sberbusiness/triplex/components/DesignTokens/types/DesignTokenTypes';

// Название токенов компонента Checkbox.
export const designTokensComponentsCheckboxKeys = [
    'Background_Default',
    'Background_Checked_Default',
    'Background_Checked_Hover',
    'Background_Disabled',

    'BorderColor_Checked',
    'BorderColor_Default',
    'BorderColor_Disabled',
    'BorderColor_Focus',
    'BorderColor_Hover',

    'Checkmark_Fill_Default',
    'Checkmark_Fill_Disabled',

    'Color_Default',
    'Color_Disabled',
] as const;
// Тип, содержащий названия токенов компонента Checkbox.
export type TDesignTokensComponentsCheckboxKeys = (typeof designTokensComponentsCheckboxKeys)[number];
// Тип, содержащий названия токенов компонента Checkbox и их значения.
export type TDesignTokensComponentsCheckboxValue = Record<TDesignTokensComponentsCheckboxKeys, TDesignTokenValue>;
// Тип, содержащий названия токенов компонента Checkbox и их значения в светлой и темной теме.
export type TDesignTokensComponentsCheckboxValues = Record<TDesignTokensComponentsCheckboxKeys, TDesignTokenValues>;
// Тип локальных токенов компонента Checkbox.
export type TDesignTokensComponentsCheckbox = {Checkbox: TDesignTokensComponentsCheckboxValue};

// Токены компонента Checkbox в светлой и темной темах.
export const Checkbox_Tokens: TDesignTokensComponentsCheckboxValues = {
    Background_Checked_Default: [{ref: 'Primary.700'}, {ref: 'Primary.500'}], // var(--triplex-Checkbox-Background_Checked_Default)
    Background_Checked_Hover: [{ref: 'Primary.500'}, {ref: 'Primary.700'}], // var(--triplex-Checkbox-Background_Checked_Hover)
    Background_Default: [{ref: 'Basic.100'}, {ref: 'Basic.0'}], // var(--triplex-Checkbox-Background_Default)
    Background_Disabled: [{ref: 'Neutral.100'}, {ref: 'Basic.500'}], // var(--triplex-Checkbox-Background_Disabled)

    BorderColor_Checked: [{ref: 'Primary.700'}, {ref: 'Primary.500'}], // var(--triplex-Checkbox-BorderColor_Checked)
    BorderColor_Default: [{ref: 'Neutral.500'}, {ref: 'Basic.300'}], // var(--triplex-Checkbox-BorderColor_Default)
    BorderColor_Disabled: [{ref: 'Neutral.500'}, {ref: 'Basic.300'}], // var(--triplex-Checkbox-BorderColor_Disabled)
    BorderColor_Focus: [{ref: 'Accent.500'}, {ref: 'Accent.500'}], // var(--triplex-Checkbox-BorderColor_Focus)
    BorderColor_Hover: [{ref: 'Primary.500'}, {ref: 'Primary.700'}], // var(--triplex-Checkbox-BorderColor_Hover)

    Checkmark_Fill_Default: [{ref: 'Basic.100'}, {ref: 'Neutral.300'}], // var(--triplex-Checkbox-Checkmark_Fill_Default)
    Checkmark_Fill_Disabled: [{ref: 'Neutral.500'}, {ref: 'Basic.300'}], // var(--triplex-Checkbox-Checkmark_Fill_Disabled)

    Color_Default: [{ref: 'Basic.700'}, {ref: 'Neutral.100'}], // var(--triplex-Checkbox-Color_Default)
    Color_Disabled: [{ref: 'Basic.700'}, {ref: 'Basic.300'}], // var(--triplex-Checkbox-Color_Disabled)
};
