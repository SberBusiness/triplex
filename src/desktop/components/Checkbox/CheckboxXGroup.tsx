import * as React from 'react';
import {TIndentSize} from '@sberbusiness/triplex/desktop/common/consts/IndentConst';
import {cssClass} from '@sberbusiness/triplex/desktop/utils/cssClass';
import {classnames} from '@sberbusiness/triplex/common/utils/classnames/classnames';

/** Свойства компонента CheckboxXGroup. */
export interface ICheckboxXGroupProps extends React.HTMLAttributes<HTMLDivElement> {
    /** Размер отступа. */
    indent?: TIndentSize;
}

/** Группа чекбоксов с направлением по оси X. */
export const CheckboxXGroup: React.FC<ICheckboxXGroupProps> = (props) => {
    const {children, className, indent = 12, ...rest} = props;
    const classNames = classnames('cssClass[checkboxXGroup]', cssClass(`indent-${indent}`), className);

    return (
        <div className={classNames} role="group" {...rest}>
            {children}
        </div>
    );
};

CheckboxXGroup.displayName = 'CheckboxXGroup';
