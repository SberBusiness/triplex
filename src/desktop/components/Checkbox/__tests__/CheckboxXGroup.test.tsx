import React from 'react';
import renderer from 'react-test-renderer';
import {allure} from '@jest/unit/allure-report';
import {CheckboxXGroup} from '@sberbusiness/triplex/desktop/components/Checkbox/CheckboxXGroup';
import {Checkbox} from '@sberbusiness/triplex/desktop/components/Checkbox/Checkbox';

jest.mock('@sberbusiness/icons/CheckboxtickStsIcon16', () => ({
    CheckboxtickStsIcon16: 'svg',
}));

jest.mock('@sberbusiness/icons/CheckboxbulkStsIcon16', () => ({
    CheckboxbulkStsIcon16: 'svg',
}));

describe('CheckboxXGroup', () => {
    beforeEach(() => {
        allure.feature('CheckboxXGroup');
    });

    it('renders correctly', () => {
        const tree = renderer
            .create(
                <CheckboxXGroup>
                    <Checkbox>Sample Text</Checkbox>
                    <Checkbox>Sample Text</Checkbox>
                    <Checkbox>Sample Text</Checkbox>
                </CheckboxXGroup>
            )
            .toJSON();
        expect(tree).toMatchSnapshot();
    });
});
