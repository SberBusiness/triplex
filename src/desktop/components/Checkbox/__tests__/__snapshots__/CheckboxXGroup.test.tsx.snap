// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CheckboxXGroup renders correctly 1`] = `
<div
  className="cssClass[checkboxXGroup] cssClass[indent-12]"
  role="group"
>
  <label
    className="cssClass[b1] cssClass[general] cssClass[regular] cssClass[extra] cssClass[label] cssClass[enabled] cssClass[nonempty]"
    onClick={[Function]}
    onMouseDown={[Function]}
  >
    <input
      className="cssClass[checkbox]"
      onBlur={[Function]}
      onFocus={[Function]}
      type="checkbox"
    />
    <span
      className="cssClass[checkboxIcon]"
    />
    <svg
      className="cssClass[checkmarkIcon]"
    />
    Sample Text
  </label>
  <label
    className="cssClass[b1] cssClass[general] cssClass[regular] cssClass[extra] cssClass[label] cssClass[enabled] cssClass[nonempty]"
    onClick={[Function]}
    onMouseDown={[Function]}
  >
    <input
      className="cssClass[checkbox]"
      onBlur={[Function]}
      onFocus={[Function]}
      type="checkbox"
    />
    <span
      className="cssClass[checkboxIcon]"
    />
    <svg
      className="cssClass[checkmarkIcon]"
    />
    Sample Text
  </label>
  <label
    className="cssClass[b1] cssClass[general] cssClass[regular] cssClass[extra] cssClass[label] cssClass[enabled] cssClass[nonempty]"
    onClick={[Function]}
    onMouseDown={[Function]}
  >
    <input
      className="cssClass[checkbox]"
      onBlur={[Function]}
      onFocus={[Function]}
      type="checkbox"
    />
    <span
      className="cssClass[checkboxIcon]"
    />
    <svg
      className="cssClass[checkmarkIcon]"
    />
    Sample Text
  </label>
</div>
`;
